metadata/src/test/java/org/apache/kafka/timeline/timeline/SnapshottableHashTableTest.java

```diff
@@ -0,0 +1,233 @@
+        registry.createSnapshot(0);
+        Iterator<TestElement> iter = table.snapshottableIterator(0);
+        assertTrue(table.snapshottableAddUnlessPresent(E_2A));
+        assertTrue(table.snapshottableAddUnlessPresent(E_3A));
+        assertIteratorContains(iter, E_1A);
+    }
+
+    @Test
+    public void testIterateOverSnapshotWhileDeletingAndReplacing() {
+        SnapshotRegistry registry = new SnapshotRegistry(0);
+        SnapshottableHashTable<TestElement> table =
+            new SnapshottableHashTable<>(registry, 1);
+        assertEquals(null, table.snapshottableAddOrReplace(E_1A));
+        assertEquals(null, table.snapshottableAddOrReplace(E_2A));
+        assertEquals(null, table.snapshottableAddOrReplace(E_3A));
+        assertEquals(E_1A, table.snapshottableRemove(E_1A));
+        assertEquals(null, table.snapshottableAddOrReplace(E_1B));
+        registry.createSnapshot(0);
+        Iterator<TestElement> iter = table.snapshottableIterator(0);
+        List<TestElement> iterElements = new ArrayList<>();
+        iterElements.add(iter.next());
+        assertEquals(E_2A, table.snapshottableRemove(E_2A));
+        assertEquals(E_3A, table.snapshottableAddOrReplace(E_3B));
+        iterElements.add(iter.next());
+        assertEquals(E_1B, table.snapshottableRemove(E_1B));
+        iterElements.add(iter.next());
+        assertFalse(iter.hasNext());
+        assertIteratorContains(iterElements.iterator(), E_1B, E_2A, E_3A);
+    }
+
+    @Test
+    public void testRevert() {
+        SnapshotRegistry registry = new SnapshotRegistry(0);
+        SnapshottableHashTable<TestElement> table =
+            new SnapshottableHashTable<>(registry, 1);
+        assertEquals(null, table.snapshottableAddOrReplace(E_1A));
+        assertEquals(null, table.snapshottableAddOrReplace(E_2A));
+        assertEquals(null, table.snapshottableAddOrReplace(E_3A));
+        registry.createSnapshot(0);
+        assertEquals(E_1A, table.snapshottableAddOrReplace(E_1B));
+        assertEquals(E_3A, table.snapshottableAddOrReplace(E_3B));
+        registry.createSnapshot(1);
+        assertEquals(3, table.snapshottableSize(Long.MAX_VALUE));
+        assertIteratorContains(table.snapshottableIterator(Long.MAX_VALUE), E_1B, E_2A, E_3B);
+        table.snapshottableRemove(E_1B);
+        table.snapshottableRemove(E_2A);
+        table.snapshottableRemove(E_3B);
+        assertEquals(0, table.snapshottableSize(Long.MAX_VALUE));
+        assertEquals(3, table.snapshottableSize(0));
+        assertEquals(3, table.snapshottableSize(1));
+        registry.revertToSnapshot(0);
+        assertIteratorContains(table.snapshottableIterator(Long.MAX_VALUE), E_1A, E_2A, E_3A);
+    }
+
+    /**
+     * Assert that the given iterator contains the given elements, in any order.
+     * We compare using reference equality here, rather than object equality.
+     */
+    private static void assertIteratorContains(Iterator<? extends Object> iter,
```

Comment:
This method tests exact set equal, not just containment. So, perhaps the method can be named more accurately.